workflows/rnaseq/downstream/helpers.Rmd

```diff
@@ -357,6 +358,64 @@ DESeqDataSetFromSalmon <- function (sampleTable, directory='.', design,
     return(object)
 }
 
+#' Make a list of dds objects
+#'
+#' This function takes a list of sample, design pairs, plus other info, to output
+#' a list of DESeq object. 
+#'
+#' @param deseq_obj_list A named list of lists. Each list should have two to three elements:
+#'       1 should be a sampletable, 2 should be a design. The 3rd is an optional list of arguments
+#'       to pass to the DESeqDataSet constructor.
+#' @param salmon.files txi object from tximport if transcript-level counts desired; NULL forces featurecounts; 
+#'        shared by all dds.
+#' @param combine.by The column to collapse technical replicates by.
+#' @param remove.version Boolean to decide whether to remove version information 
+#' @return A list of dds objects.
+#' 
+make.dds.list <- function(deseq_obj_list, salmon.files=NULL, 
+                          combine.by=FALSE, remove.version=TRUE, ...){
+    make.dds <- function(design_data, salmon.files, combine.by, remove.version, ...)
+    {
+        
+        subset.counts <- FALSE
+        colData <- pluck(design_data, 1)
+        design <- pluck(design_data, 2)
+        # arg list passes args to either the CombinedFeatureCounts or FromSalmon constructor
+        arg_list <- pluck(design_data, 3)
+
+        if(is.null(salmon.files)){
+            # Load gene-level counts
+           dds<- exec(DESeqDataSetFromCombinedFeatureCounts,
+                '../data/rnaseq_aggregation/featurecounts.txt',
+                sampletable=colData,
+                # NOTE: What design to use?------------------------------------------------
+                #   This object will be used for EDA, so it's recommended to use a single
+                #   factor that describes the samples. We'll be recreating another dds
+                #   later for the actual contrasts of interest.
+                design=design,
+                !!!arg_list)
```

Comment:
I've never seen `!!!` before -- it looks to be an equivalent of Python's `**kwargs`, is that right?

Reply:
It is very similar. But there are some differences related to named and unnamed parameters. Also, in this case, the following function doesn't have tidy dots, so the exec is also required so the !!! evaluates first. Hadley once again has our back on this [here](https://adv-r.hadley.nz/quasiquotation.html). I actually wanted to put together a few resources I found on the subject. I'll put it in a wiki.